### PR TITLE
Add tests to the current securitas codebase

### PR DESCRIPTION
**Pull request by abompard**
Use WTForm for the login form and add tests on the login and logout controllers.

### CODE REVIEW

File: securitas/controller/authentication.py [hunk 1]
```diff
@@ -1,43 +1,55 @@
...
+from securitas.form.login_user import LoginUserForm
+from securitas.security.ipa import maybe_ipa_login, maybe_ipa_session
 
 
 @app.route('/logout')
-@with_ipa(app, session)
-def logout(ipa):
+def logout():
+    """Log the user out."""
+    # Don't use the with_ipa() decorator, otherwise anonymous users visiting this endpoint will be
```
Comment (relrod) [inline securitas/controller/authentication.py:12]: Heh, fair. I guess I didn't consider that an issue because an anonymous user shouldn't be going to this endpoint. :laughing:

File: securitas/templates/login.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{% extends "with_navbar.html" %}
+{% block title %}Self-Service Portal{% endblock %}
```
Comment (relrod) [inline securitas/templates/login.html]: Perhaps use a more descriptive title that mentions "login" somewhere?

File: securitas/form/login_user.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from flask_wtf import FlaskForm
+from wtforms import StringField, PasswordField
+from wtforms.validators import DataRequired
+
+
+class LoginUserForm(FlaskForm):
```
Comment (relrod) [inline securitas/form/login_user.py:6]: Is it worth trying to encode some rules here about what constitutes a valid username/password? We'll basically be re-encoding the IPA rules I guess, but the advantage is that some of them could be verified on the client side before the form is ever submitted (e.g. min/max length).

File: securitas/controller/authentication.py [hunk 4]
```diff
@@ -1,43 +1,55 @@
...
+            ipa = maybe_ipa_login(app, session, username, password)
+        except python_freeipa.exceptions.PasswordExpired:
+            flash('Password expired. Please reset it.', 'red')
+            return redirect(url_for('password_reset'))
+        except python_freeipa.exceptions.Unauthorized as e:
+            form.errors['non_field_errors'] = [e.message]
+        except python_freeipa.exceptions.FreeIPAError as e:
+            # If we made it here, we hit something weird not caught above. We didn't
+            # bomb out, but we don't have IPA creds, either. Boot us back to /.
+            app.logger.error(
```
Comment (relrod) [inline securitas/controller/authentication.py:38]: I *think* this is okay because the IPA server should never include passwords in the response, but we should be mindful to make sure nothing we log will ever include a password.

**Review by relrod — COMMENTED** [review]
Some inlines but overall this looks fine to me.

File: securitas/form/login_user.py (hunk 3 above)
Comment (abompard) [reply securitas/form/login_user.py:6]: Yeah I asked myself that question (for the registration form), and I thought that if we want to change the password policy we'd have to do it in two different places, which is error-prone.
So I've only left basic "not-empty" validators.

File: securitas/controller/authentication.py (hunk 4 above)
Comment (abompard) [reply securitas/controller/authentication.py:38]: Oh good point, I didn't think of that. Hopefully they don't put passwords in the `message` attribute of their exceptions, though.

File: securitas/templates/login.html (hunk 2 above)
Comment (abompard) [reply securitas/templates/login.html]: Will do.

File: securitas/tests/unit/conftest.py [hunk 5]
```diff
@@ -52,18 +52,49 @@ def vcr_cassette_dir(request):
...
+            user_password=password,
             login_shell='/bin/bash',
+            fascreationtime=f"{now.isoformat()}Z",
         )
         ipa = untouched_ipa_client(app)
-        ipa.change_password('dummy', 'dummy_password', 'dummy_password')
-        yield
-    finally:
-        ipa_admin.user_del('dummy')
+        ipa.change_password(name, password, password)
```
Comment (StephenCoady) [inline securitas/tests/unit/conftest.py:71]: Question - why is this line necessary? Is it a temporary password if created using the API?

File: securitas/tests/unit/controller/test_authentication.py [hunk 6]
```diff
@@ -1,71 +1,171 @@
...
+# Login
 
 
 @pytest.mark.vcr()
 def test_login(client):
     """Test a successful Login"""
     result = client.post(
         '/login',
-        data={"username": "dudemcpants", "password": "n:nPv{P].9}]!q$RE%w<38@",},
+        data={"username": "dudemcpants", "password": "n:nPv{P].9}]!q$RE%w<38@"},
```
Comment (StephenCoady) [inline securitas/tests/unit/controller/test_authentication.py]: shouldn't this be the dummy user now? the cassettes might be hiding this failing.

Comment (abompard) [reply securitas/tests/unit/controller/test_authentication.py]: Correct, it's only working because we didn't update the cassettes. Also, it should use the `dummy_user` fixture.

File: securitas/tests/unit/conftest.py (hunk 5 above)
Comment (abompard) [reply securitas/tests/unit/conftest.py:71]: When a user is create with the API, it's created with an expired password, so they are forced to change it before they can login. We don't want that, so we're changing the password ourselves. It's also what the registration controller does.

**Review by StephenCoady — APPROVED** [review]
